keydata.go

```diff
@@ -0,0 +1,402 @@
+	}
+
+	h = crypto.SHA256.New()
+	if err := tpm2.MarshalToWriter(h, &policyUpdateData.Data); err != nil {
+		panic(fmt.Sprintf("cannot marshal dynamic authorization policy update data: %v", err))
+	}
+
+	if !bytes.Equal(h.Sum(nil), policyUpdateData.CreationData.OutsideInfo) {
+		return nil, keyFileError{errors.New("key data file and dynamic authorization policy update data file mismatch: digest doesn't match creation data")}
+	}
+
+	authKey, err := x509.ParsePKCS1PrivateKey(policyUpdateData.Data.AuthKey)
+	if err != nil {
+		return nil, keyFileError{xerrors.Errorf("cannot parse dynamic authorization policy signing key: %w", err)}
+	}
+
+	authPublicKey := rsa.PublicKey{
+		N: new(big.Int).SetBytes(data.StaticPolicyData.AuthPublicKey.Unique.RSA()),
+		E: int(data.StaticPolicyData.AuthPublicKey.Params.RSADetail().Exponent)}
+	if authKey.PublicKey.E != authPublicKey.E || authKey.PublicKey.N.Cmp(authPublicKey.N) != 0 {
+		return nil, keyFileError{errors.New("dynamic authorization policy signing private key doesn't match public key")}
+	}
+
+	return pinIndexPublic, nil
+}
+
+// readAndValidateKeyData will deserialize keyData and keyPolicyUpdateData from the provided io.Readers and then perform some correctness
+// checking. On success, it returns the keyData, keyPolicyUpdateData and the validated public area of the PIN NV index.
+func readAndValidateKeyData(tpm *tpm2.TPMContext, keyFile, keyPolicyUpdateFile io.Reader, session tpm2.SessionContext) (*keyData, *keyPolicyUpdateData, *tpm2.NVPublic, error) {
+	// Read the key data
+	data, err := readKeyData(keyFile)
+	if err != nil {
+		return nil, nil, nil, xerrors.Errorf("cannot read key data: %w", err)
+	}
+
+	var policyUpdateData *keyPolicyUpdateData
+	if keyPolicyUpdateFile != nil {
+		var err error
+		policyUpdateData, err = readKeyPolicyUpdateData(keyPolicyUpdateFile)
+		if err != nil {
+			return nil, nil, nil, xerrors.Errorf("cannot read dynamic policy authorization data: %w", err)
+		}
+	}
+
+	pinNVPublic, err := validateKeyData(tpm, data, policyUpdateData, session)
+	if err != nil {
+		return nil, nil, nil, xerrors.Errorf("key data validation failed: %w", err)
+	}
+
+	return data, policyUpdateData, pinNVPublic, nil
+}
+
+// SealedKeyObject corresponds to a sealed key data file and exists to provide access to some read only operations on the underlying
+// file without having to read and deserialize the key data file more than once.
+type SealedKeyObject struct {
+	data *keyData
+}
+
+// AuthMode2F indicates the 2nd-factor authentication type for this sealed key object.
+func (k *SealedKeyObject) AuthMode2F() AuthMode {
+	return k.data.AuthModeHint
+}
+
+// PINIndexHandle indicates the handle of the NV index used for PIN support for this sealed key object.
+func (k *SealedKeyObject) PINIndexHandle() tpm2.Handle {
+	return k.data.StaticPolicyData.PinIndexHandle
+}
+
+// LoadSealedKeyObject loads a sealed key data file created by SealKeyToTPM from the specified path. If the file cannot be opened,
```

Comment:
this actually loads data in the tpm only temporarily in validateKeyData?  should it be called ReadSealedObject instead ?

Reply:
I've renamed this to ReadSealedKeyObject. This function actually doesn't load anything in to the TPM. The validateKeyData function is only called from functions that aren't called during in early boot, as a way to perform additional diagnostic checks.